File: stacks/nodejs/react-fastify/rest/bff/src/flow/gitlab/adapter.test.ts
import assert from "node:assert/strict";
import fs from "node:fs";
import path from "node:path";
import { test } from "node:test";
import YAML from "yaml";
import type { ProviderAdapterInput } from "../types";
import { buildGitLabProviderInput } from "./adapter";
import {
  agingImplementationSource,
  blockedOnReviewSelfManagedSource,
  blockedOnReviewSource,
  riskAggregationSource,
  trunkIntegrationFailureSource,
  unclearOwnershipSource,
  waitingOnEvidenceSource,
} from "./fixtures";

function findRepoRoot(startDir: string): string {
  let current = startDir;
  while (true) {
    if (fs.existsSync(path.join(current, "schema"))) {
      return current;
    }
    const parent = path.dirname(current);
    if (parent === current) {
      throw new Error("Repository root not found from adapter tests");
    }
    current = parent;
  }
}

const repoRoot = findRepoRoot(__dirname);
const fixtureDir = path.join(repoRoot, "schema", "fixtures", "provider-adapter-input");

function loadNormalizedFixture(filename: string): ProviderAdapterInput {
  const filePath = path.join(fixtureDir, filename);
  const content = fs.readFileSync(filePath, "utf8");
  const parsed = YAML.parse(content) as Partial<ProviderAdapterInput> & Record<string, unknown>;
  return {
    repository: parsed.repository ?? { provider: "gitlab", provider_id: "", full_name: "", default_branch: "", fetched_at: "" },
    changes: parsed.changes ?? [],
    actors: parsed.actors ?? [],
    review_states: parsed.review_states ?? [],
    validation_runs: parsed.validation_runs ?? [],
    merge_events: parsed.merge_events ?? [],
    evidence_states: parsed.evidence_states ?? [],
    ownership_hints: parsed.ownership_hints ?? [],
  };
}

function pruneUndefined<T>(value: T): T {
  if (Array.isArray(value)) {
    return value.map((item) => pruneUndefined(item)) as T;
  }
  if (value && typeof value === "object") {
    const entries = Object.entries(value as Record<string, unknown>).filter(
      ([, v]) => v !== undefined,
    );
    const result: Record<string, unknown> = {};
    for (const [key, val] of entries) {
      result[key] = pruneUndefined(val);
    }
    return result as T;
  }
  return value;
}

function normalizeOutput(input: ProviderAdapterInput): ProviderAdapterInput {
  const normalized = pruneUndefined<ProviderAdapterInput>({
    ...input,
    changes: [...input.changes].sort((a, b) => a.provider_id.localeCompare(b.provider_id)),
    actors: [...input.actors].sort((a, b) => a.provider_id.localeCompare(b.provider_id)),
    review_states: [...input.review_states].sort((a, b) => a.change_id.localeCompare(b.change_id)),
    validation_runs: [...input.validation_runs].sort((a, b) => {
      if (a.change_id !== b.change_id) {
        return a.change_id.localeCompare(b.change_id);
      }
      if (a.scope !== b.scope) {
        return a.scope.localeCompare(b.scope);
      }
      return (a.run_at ?? "").localeCompare(b.run_at ?? "");
    }),
    merge_events: [...input.merge_events].sort((a, b) => a.change_id.localeCompare(b.change_id)),
    evidence_states: [...(input.evidence_states ?? [])].sort((a, b) => {
      if (a.change_id !== b.change_id) {
        return a.change_id.localeCompare(b.change_id);
      }
      return (a.as_of ?? "").localeCompare(b.as_of ?? "");
    }),
    ownership_hints: [...input.ownership_hints].sort((a, b) => {
        const pathA = a.path_pattern ?? "";
        const pathB = b.path_pattern ?? "";
        if (pathA !== pathB) {
          return pathA.localeCompare(pathB);
        }
        return (a.source ?? "").localeCompare(b.source ?? "");
      }),
  });

  return normalized;
}

test("normalizes GitLab data for blocked-on-review scenario (SaaS)", () => {
  const result = buildGitLabProviderInput(blockedOnReviewSource);
  const expected = loadNormalizedFixture("blocked-on-review-gitlab.yaml");

  assert.deepStrictEqual(normalizeOutput(result), normalizeOutput(expected));
});

test("normalizes GitLab data for blocked-on-review self-managed scenario", () => {
  const result = buildGitLabProviderInput(blockedOnReviewSelfManagedSource);
  const expected = loadNormalizedFixture("blocked-on-review-gitlab-self-managed.yaml");

  assert.deepStrictEqual(normalizeOutput(result), normalizeOutput(expected));
});

test("normalizes GitLab data for trunk integration failure scenario", () => {
  const result = buildGitLabProviderInput(trunkIntegrationFailureSource);
  const expected = loadNormalizedFixture("trunk-integration-failed-gitlab.yaml");

  assert.deepStrictEqual(normalizeOutput(result), normalizeOutput(expected));
});

test("normalizes GitLab data for unclear ownership scenario", () => {
  const result = buildGitLabProviderInput(unclearOwnershipSource);
  const expected = loadNormalizedFixture("unclear-ownership-gitlab.yaml");

  assert.deepStrictEqual(normalizeOutput(result), normalizeOutput(expected));
});

test("normalizes GitLab data for waiting on evidence scenario", () => {
  const result = buildGitLabProviderInput(waitingOnEvidenceSource);
  const expected = loadNormalizedFixture("waiting-on-evidence-gitlab.yaml");

  assert.deepStrictEqual(normalizeOutput(result), normalizeOutput(expected));
});

test("normalizes GitLab data for aging implementation scenario", () => {
  const result = buildGitLabProviderInput(agingImplementationSource);
  const expected = loadNormalizedFixture("aging-implementation-gitlab.yaml");

  assert.deepStrictEqual(normalizeOutput(result), normalizeOutput(expected));
});

test("normalizes GitLab data for risk aggregation scenario", () => {
  const result = buildGitLabProviderInput(riskAggregationSource);
  const expected = loadNormalizedFixture("risk-aggregation-gitlab.yaml");

  assert.deepStrictEqual(normalizeOutput(result), normalizeOutput(expected));
});

File: stacks/go/net-http/rest/bff/flow/gitlab/adapter_test.go
package gitlab

import (
	"os"
	"path/filepath"
	"reflect"
	"runtime"
	"sort"
	"testing"

	"gopkg.in/yaml.v3"

	"idp-go-net-http-rest/bff/flow"
)

func loadFixture(t *testing.T, name string) flow.ProviderAdapterInput {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatalf("unable to resolve caller")
	}
	base := filepath.Dir(file)
	// stacks/go/net-http/rest/bff/flow/gitlab -> repo root is seven levels up.
	path := filepath.Join(base, "..", "..", "..", "..", "..", "..", "..", "schema", "fixtures", "provider-adapter-input", name+".yaml")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture %s: %v", name, err)
	}
	var input flow.ProviderAdapterInput
	if err := yaml.Unmarshal(data, &input); err != nil {
		t.Fatalf("unmarshal fixture %s: %v", name, err)
	}
	if input.EvidenceStates == nil {
		input.EvidenceStates = []flow.NormalizedEvidenceState{}
	}
	if input.MergeEvents == nil {
		input.MergeEvents = []flow.NormalizedMergeEvent{}
	}
	return input
}

func sortInput(input flow.ProviderAdapterInput) flow.ProviderAdapterInput {
	out := input
	out.Changes = append([]flow.NormalizedChange(nil), input.Changes...)
	sort.SliceStable(out.Changes, func(i, j int) bool { return out.Changes[i].ProviderID < out.Changes[j].ProviderID })

	out.Actors = append([]flow.NormalizedActor(nil), input.Actors...)
	sort.SliceStable(out.Actors, func(i, j int) bool { return out.Actors[i].ProviderID < out.Actors[j].ProviderID })

	out.ReviewStates = append([]flow.NormalizedReviewState(nil), input.ReviewStates...)
	sort.SliceStable(out.ReviewStates, func(i, j int) bool { return out.ReviewStates[i].ChangeID < out.ReviewStates[j].ChangeID })

	out.ValidationRuns = append([]flow.NormalizedValidationRun(nil), input.ValidationRuns...)
	sort.SliceStable(out.ValidationRuns, func(i, j int) bool {
		a := out.ValidationRuns[i]
		b := out.ValidationRuns[j]
		if a.ChangeID != b.ChangeID {
			return a.ChangeID < b.ChangeID
		}
		if a.Scope != b.Scope {
			return a.Scope < b.Scope
		}
		return a.RunAt < b.RunAt
	})

	out.MergeEvents = append([]flow.NormalizedMergeEvent(nil), input.MergeEvents...)
	sort.SliceStable(out.MergeEvents, func(i, j int) bool { return out.MergeEvents[i].ChangeID < out.MergeEvents[j].ChangeID })

	out.OwnershipHints = append([]flow.NormalizedOwnershipHint(nil), input.OwnershipHints...)
	sort.SliceStable(out.OwnershipHints, func(i, j int) bool {
		a := out.OwnershipHints[i]
		b := out.OwnershipHints[j]
		if a.PathPattern != b.PathPattern {
			return a.PathPattern < b.PathPattern
		}
		return a.Source < b.Source
	})

	out.EvidenceStates = append([]flow.NormalizedEvidenceState(nil), input.EvidenceStates...)
	sort.SliceStable(out.EvidenceStates, func(i, j int) bool {
		a := out.EvidenceStates[i]
		b := out.EvidenceStates[j]
		if a.ChangeID != b.ChangeID {
			return a.ChangeID < b.ChangeID
		}
		return a.AsOf < b.AsOf
	})

	for i := range out.Changes {
		if out.Changes[i].WorkItemRefs == nil {
			out.Changes[i].WorkItemRefs = []flow.NormalizedWorkItemRef{}
		}
	}

	for i := range out.OwnershipHints {
		if out.OwnershipHints[i].OwnerActorIDs == nil {
			out.OwnershipHints[i].OwnerActorIDs = []string{}
		}
		if out.OwnershipHints[i].OwnerTeamNames == nil {
			out.OwnershipHints[i].OwnerTeamNames = []string{}
		}
	}

	if out.MergeEvents == nil {
		out.MergeEvents = []flow.NormalizedMergeEvent{}
	}
	if out.EvidenceStates == nil {
		out.EvidenceStates = []flow.NormalizedEvidenceState{}
	}
	return out
}

func assertProviderInputEqual(t *testing.T, got, want flow.ProviderAdapterInput) {
	t.Helper()
	gotS := sortInput(got)
	wantS := sortInput(want)
	if !reflect.DeepEqual(gotS, wantS) {
		t.Fatalf("provider input mismatch\n got: %#v\nwant: %#v", gotS, wantS)
	}
}

func TestGitLabAdapterBlockedOnReview(t *testing.T) {
	got := BuildProviderInput(blockedOnReviewSource())
	want := loadFixture(t, "blocked-on-review-gitlab")
	assertProviderInputEqual(t, got, want)
}

func TestGitLabAdapterBlockedOnReviewSelfManaged(t *testing.T) {
	got := BuildProviderInput(blockedOnReviewSelfManagedSource())
	want := loadFixture(t, "blocked-on-review-gitlab-self-managed")
	assertProviderInputEqual(t, got, want)
}

func TestGitLabAdapterTrunkIntegrationFailure(t *testing.T) {
	got := BuildProviderInput(trunkIntegrationFailureSource())
	want := loadFixture(t, "trunk-integration-failed-gitlab")
	assertProviderInputEqual(t, got, want)
}

func TestGitLabAdapterUnclearOwnership(t *testing.T) {
	got := BuildProviderInput(unclearOwnershipSource())
	want := loadFixture(t, "unclear-ownership-gitlab")
	assertProviderInputEqual(t, got, want)
}

func TestGitLabAdapterWaitingOnEvidence(t *testing.T) {
	got := BuildProviderInput(waitingOnEvidenceSource())
	want := loadFixture(t, "waiting-on-evidence-gitlab")
	assertProviderInputEqual(t, got, want)
}

func TestGitLabAdapterAgingImplementation(t *testing.T) {
	got := BuildProviderInput(agingImplementationSource())
	want := loadFixture(t, "aging-implementation-gitlab")
	assertProviderInputEqual(t, got, want)
}

func TestGitLabAdapterRiskAggregation(t *testing.T) {
	got := BuildProviderInput(riskAggregationSource())
	want := loadFixture(t, "risk-aggregation-gitlab")
	assertProviderInputEqual(t, got, want)
}
